File: 프로그래머스/lv2/92341. 주차 요금 계산/주차 요금 계산.cpp
#include <string>
#include <vector>
#include <sstream>
#include <map>
#include <iterator>
#include <iostream>
#include <algorithm>

#define X first
#define Y second

using namespace std;

vector<string> split(string input, char delimeter){
    vector<string> answer;
    stringstream ss(input);
    string temp;
    
    while(getline(ss, temp, delimeter)){
        answer.push_back(temp);
    }
    
    return answer;
}

int time_to_fee(int time, vector<int> fees){
    if(time <= fees[0]){
        return fees[1];
    }
    else{
        time -= fees[0];
        if(time % fees[2] == 0) time /= fees[2];
        else time = time / fees[2] + 1;
        return fees[1] + time * fees[3];
    }
}

vector<int> solution(vector<int> fees, vector<string> records) {
    vector<int> answer;
    vector<pair<int, int> >record;
    for(int i = 0; i < records.size(); i++){
        vector<string> temp;
        vector<string> time;
        temp = split(records[i], ' ');
        time = split(temp[0], ':');
        int int_time = stoi(time[0]) * 60 + stoi(time[1]);
        record.push_back(make_pair(stoi(temp[1]), int_time));
    }
    sort(record.begin(), record.end());
    for(int i = 0; i < record.size(); i++){
        cout << record[i].X << ' ' << record[i].Y << '\n';
    }
    int cnt = 1;
    int value = 0, last_time = 23*60+59, money = 0;
    for(int i = 0; i < record.size() - 1; i++){
        if(record[i].X == record[i+1].X && cnt == 1){
            value += record[i+1].Y - record[i].Y;
            cnt = 0;
            if(i == record.size() - 2){
                money = time_to_fee(value, fees);
                answer.push_back(money);
            }
        }
        else if(record[i].X == record[i+1].X && cnt == 0){
            cnt = 1;
        }
        else if(record[i].X != record[i+1].X && cnt == 1){
            value += last_time - record[i].Y;
            money = time_to_fee(value, fees);
            answer.push_back(money);
            value = 0;
        }
        else{
            money = time_to_fee(value, fees);
            answer.push_back(money);
            value = 0;
            cnt = 1;
        }
        cout << money << '\n';
    }
    if(cnt == 1){
        value += last_time - record[record.size() - 1].Y;
        money = time_to_fee(value, fees);
        answer.push_back(money);
    }
    return answer;
}

File: 프로그래머스/lv2/92341. 주차 요금 계산/주차 요금 계산.py
def cal_fee(fees, time):
    if time <= fees[0]:
        return fees[1]
    else:
        mok = (time - fees[0]) // fees[2]
        if (time - fees[0]) % fees[2] != 0:
            mok += 1
        return fees[1] + mok * fees[3]

def cal_time(re_str):
    temp = re_str.split(":")
    return int(temp[0]) * 60 + int(temp[1])

def solution(fees, records):
    record = [rec.split() for rec in records]
    print(record)
    record = sorted(record, key=lambda record: record[1])
    print(record)
    car_number = dict()
    fee = dict()
    for re in record:
        if re[1] not in car_number:
            car_number[re[1]] = cal_time(re[0])
        else:
            past = car_number.pop(re[1])
            now = cal_time(re[0])
            if re[1] not in fee:
                fee[re[1]] = now - past
            else:
                fee[re[1]] += now - past
    for cn in car_number:
        now = 1439
        if cn not in fee:
            fee[cn] = now - car_number[cn]
        else:
            fee[cn] += now - car_number[cn]
    for fe in fee:
        fee[fe] = cal_fee(fees, fee[fe])
    answer = list(map(lambda x:x[1], sorted(fee.items())))
    return answer